utils/s3/client.cc

```diff
@@ -234,26 +235,92 @@ future<> map_s3_client_exception(std::exception_ptr ex) {
         auto e = std::current_exception();
         return make_exception_future<>(storage_io_error(EIO, format("S3 error ({})", e)));
     }
+}
+
+static future<aws::aws_error> look_for_errors(const http::reply& reply, input_stream<char>& in_, bool is_mpu_completion) {
+    if (reply._status >= http::reply::status_type::multiple_choices || is_mpu_completion) {
+        auto response_body = co_await util::read_entire_stream_contiguous(in_);
+
+        std::optional<aws::aws_error> possible_error = aws::aws_error::parse(std::move(response_body));
+        if (possible_error) {
+            co_return possible_error.value();
+        }
+        if (reply._status >= http::reply::status_type::multiple_choices) {
+            co_return aws::aws_error::from_http_code(reply._status);
+        }
+    }
+
+    co_return aws::aws_error{aws::aws_error_type::OK, aws::retryable::no};
+}
 
+future<> client::do_retryable_request(group_client& gc, http::request req, http::experimental::client::reply_handler handler) {
+    return do_with(uint32_t{0}, std::move(req), std::move(handler),
```

Comment:
can we use C++20 coroutines here? for better readability.

Reply:
Done